### PR TITLE
[SPARK-32907][ML] adaptively blockify instances - LinearSVC

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, use `blockSizeInMB` instead of `blockSize`(#rows) to control the stacking of vectors;
2, infer an appropriate `blockSizeInMB` if set 0, based on data sparsity;

### Why are the changes needed?
the performance gain is mainly related to the nnz of block.

### Does this PR introduce _any_ user-facing change?
yes, param `blockSize` -> `blockSizeInMB` in master

### How was this patch tested?
added testsuites and performance test (result attached in [ticket](https://issues.apache.org/jira/browse/SPARK-32907))

### CODE REVIEW

**Review by zero323 — COMMENTED** [review]
We should also have modify Python type annotations:

- Adjust type definitiion and `__init__`  of [`_LinearSVCParams`](https://github.com/apache/spark/blob/dfb7790a9dad8e98bd27001a613b4e13a5eb9d51/python/pyspark/ml/classification.pyi#L165-L178)
- Adjust `__init__` and `setParams` of [`LinearSVC`](https://github.com/apache/spark/blob/dfb7790a9dad8e98bd27001a613b4e13a5eb9d51/python/pyspark/ml/classification.pyi#L186-L219)
- Add annotations for new method and class.

**Comment by zhengruifeng** [comment]
also ping @WeichenXu123

**Comment by zero323** [comment]
Thanks for the changes @zhengruifeng 

Could also modify `shared.pyi` and add the import?

```
diff --git a/python/pyspark/ml/classification.pyi b/python/pyspark/ml/classification.pyi
index 74d11cfd11..241f5baf8d 100644
--- a/python/pyspark/ml/classification.pyi
+++ b/python/pyspark/ml/classification.pyi
@@ -26,6 +26,7 @@ from pyspark.ml.base import _PredictorParams
 from pyspark.ml.param.shared import (
     HasAggregationDepth,
     HasBlockSize,
+    HasBlockSizeInMB,
     HasElasticNetParam,
     HasFitIntercept,
     HasMaxIter,
diff --git a/python/pyspark/ml/param/shared.pyi b/python/pyspark/ml/param/shared.pyi
index 5999c0eaa4..bbb4890455 100644
--- a/python/pyspark/ml/param/shared.pyi
+++ b/python/pyspark/ml/param/shared.pyi
@@ -185,3 +185,8 @@ class HasBlockSize(Params):
     blockSize: Param[int]
     def __init__(self) -> None: ...
     def getBlockSize(self) -> int: ...
+
+class HasBlockSizeInMB(Params):
+    blockSizeInMB: Param[float]
+    def __init__(self) -> None: ...
+    def getBlockSizeInMB(self) -> float: ...
```

In general you can verify consistency of annotations locally (sorry, still working on the pipeline):

```bash
# Checks annotations
mypy --no-incremental --config python/mypy.ini python/pyspark
# Checks Spark examples
MYPYPATH=python/ mypy --no-incremental --config python/mypy.ini examples/src/main/python/ml examples/src/main/python/sql examples/src/main/python/sql/streaming
```

**Comment by zhengruifeng** [comment]
@zero323 Thank you so much. I involved other commits here by mistake. I will resend a new PR.

File: project/plugins.sbt [hunk 1]
```diff
@@ -27,8 +27,6 @@ addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.15.0")
 
 addSbtPlugin("com.typesafe.sbteclipse" % "sbteclipse-plugin" % "5.2.4")
 
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.9.2")
-
```
Comment (WeichenXu123) [inline project/plugins.sbt:31]: why change this ?

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -213,17 +213,23 @@ object SparkBuild extends PomBuild {
...
+    publishLocalConfiguration in SbtCompile := PublishConfiguration()
+        .withResolverName("ivyLocal")
+        .withArtifacts(packagedArtifacts.value.toVector)
+        .withLogging(ivyLoggingLevel.value),
     publishMavenStyle in MavenCompile := true,
+    publishMavenStyle in SbtCompile := false,
     publishLocal in MavenCompile := publishTask(publishLocalConfiguration in MavenCompile).value,
-    publishLocalBoth := Seq(publishLocal in MavenCompile, publishLocal).dependOn.value,
+    publishLocal in SbtCompile := publishTask(publishLocalConfiguration in SbtCompile).value,
+    publishLocal := Seq(publishLocal in MavenCompile, publishLocal in SbtCompile).dependOn.value,
```
Comment (WeichenXu123) [inline project/SparkBuild.scala:232]: why change this ?

File: python/pyspark/resource/profile.pyi [hunk 3]
```diff
@@ -22,7 +22,7 @@ from pyspark.resource.requests import (  # noqa: F401
     TaskResourceRequest as TaskResourceRequest,
     TaskResourceRequests as TaskResourceRequests,
 )
-from typing import overload, Dict, Optional, Union
```
Comment (WeichenXu123) [inline python/pyspark/resource/profile.pyi:25]: why change this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 4]
```diff
@@ -76,7 +76,7 @@ class MicroBatchExecution(
     // transformation is responsible for replacing attributes with their final values.
 
     val disabledSources =
-      sparkSession.sqlContext.conf.disabledV2StreamingMicroBatchReaders.split(",")
+      Utils.stringToSeq(sparkSession.sqlContext.conf.disabledV2StreamingMicroBatchReaders)
```
Comment (WeichenXu123) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:79]: why change this ?

**Review by WeichenXu123 — COMMENTED** [review]
This PR includes many irrelevant changes ?

**Pull request closed without merging**